### DISCUSSION TITLE
Build PyQt

### DISCUSSION THREAD

**Discussion by Mame29** [termux/termux-packages#8775] [General]
I have successfully built PyQt5 with Python3.10.2, but an error is coming out Is there
![Screenshot_20220127-085918](https://user-images.githubusercontent.com/72637907/151279073-3cf15c7c-b4d0-4acd-984e-da85aa4b6178.png)

**Comment by SDRausty** [comment]
> Saya telah berhasil membangun PyQt5 dengan Python3.10.2, tetapi ada kesalahan yg keluar
Ada kah yg bisa membantu ?

What language are you using?  Try again with translate.google.com and similar please.  

Please use the reverse translate feature.  Tap the horizontal double arrow repeatedly to get the text correct into English.  Using very short sentences can help in getting the translation correct.

**Reply by Mame29** [reply]
Ok

**Reply by SDRausty** [reply]
Thank you.

**Reply by Mame29** [reply]
So, you know the solution?

**Reply by SDRausty** [reply]
> So, you know the solution?

No, but we can try to find one.  Have you tried building PyQt with [TermuxArch](https://github.com/TermuxArch/TermuxArch) on your smartphone?

**Reply by SDRausty** [reply]
If it does not work in native Termux, I try with [TermuxArch](https://github.com/TermuxArch/TermuxArch).  Why?  Maybe it will work out of the box in proot, and many times it does work on the first try;  Just like a charm!

**Reply by SDRausty** [reply]
I just checked `yay pyqt` in  [TermuxArch](https://github.com/TermuxArch/TermuxArch), and you might not even need to build it because the package that you are searching for might already be packaged in Arch Linux:
![Screenshot_20220127-041657](https://user-images.githubusercontent.com/27742457/151329297-c48fa8da-449a-4662-beb6-41f7b37e1e0b.png)
 There are 84 Arch Linux packages available that `yay pyqt` provides in Arch Linux in Termux PRoot in your smartphone today;  Good luck!

**Reply by SDRausty** [reply]
> So, you know the solution?

You might also still be able to try with python2 in native Termux, but I am not sure how well using python2 instead of python3 solutions work in Termux today.

**Reply by Mame29** [reply]
> > Jadi, Anda tahu solusinya?
> 
> Anda mungkin juga masih dapat mencoba dengan python2 di Termux asli, tetapi saya tidak yakin seberapa baik menggunakan solusi python2 alih-alih python3 bekerja di Termux hari ini.

maybe it takes up a lot of storage

**Reply by SDRausty** [reply]
Arch Linux is not small:
```
$ f -i usrspace *h|f lt
11:maintenanceroutines.bash:235:if [[ "$USRSPACE" -lt "1500000" ]]
13:maintenanceroutines.bash:243:if [[ "$USRSPACE" -lt "1250000" ]]
15:maintenanceroutines.bash:251:if [[ "$USRSPACE" -lt "800000" ]]
17:maintenanceroutines.bash:259:if [[ "$USRSPACE" -lt "600000" ]]
```
`setupTermuxArch` should let the user know if there is not enough space on device before installing.

**Reply by SDRausty** [reply]
> might not even need to build it because the package that you are searching for might already be packaged

![Screenshot_20220127-065948](https://user-images.githubusercontent.com/27742457/151355102-378c6afe-ab00-4579-9858-2291e2357def.png)

```
$ yay pyqt
.
.
.
Total Download Size:    29.67 MiB
Total Installed Size:  179.10 MiB
```

**Reply by Mame29** [reply]
😱

**Reply by SDRausty** [reply]
Can you please elaborate on the face screamming emojii shared and how it helps this project develop and grow?

**Reply by Mame29** [reply]
is there any other way besides using proot ?, because my storage is not sufficient

**Comment by xtkoba** [comment]
Due to notorious android-ndk/ndk#201. Needs something like

```bash
export LDFLAGS+=" -lpython3.10"
```

before building.

**Reply by Mame29** [reply]
> Due to notorious [android/ndk#201](https://github.com/android/ndk/issues/201). Needs something like
> 
> ```shell
> export LDFLAGS+=" -lpython3.10"
> ```
> 
> before building.

OK is work, but now any new error
![Screenshot_20220127-175732](https://user-images.githubusercontent.com/72637907/151345776-a416e194-a1d5-44e9-aef4-0aba9006df39.png)

**Reply by xtkoba** [reply]
`_ZN13QSharedMemory16staticMetaObjectE` is defined in `lib/libQt5Core.so`. Also needs `-lQt5Core` in `LDFLAGS` I suppose.

**Reply by Mame29** [reply]
libQt5Core.so is already linked, but result is `_ZN13QSharedMemory16staticMetaObjectE`

**Reply by Mame29** [reply]
does it have something to do with `-DTERMUX_SHMEM_STUBS`

**Reply by xtkoba** [reply]
Uh-oh, in fact it does not seem to be defined anywhere.

**Reply by Mame29** [reply]
we won't know until we try :D

**Reply by xtkoba** [reply]
It has been revealed that the current package of `qt5-qtbase` does not contain the following classes:

```
QCalendar
QConcatenateTablesProxyModel
QSharedMemory
QTransposeProxyModel
```

I am not sure if this is a bug or not. There is a report that <a href="https://bugreports.qt.io/browse/QTBUG-24109">the shared memory support is disabled when cross-compiling with MinGW</a>, which I am not sure is relevant here.

A workaround would be to get rid of the references to those classes in PyQt5.

Another would be to prepare a dummy library which PyQt5 is linked against (or is `LD_PRELOAD`ed) that defines those missing symbols.

**Reply by Mame29** [reply]
> It has been revealed that the current package of `qt5-qtbase` does not contain the following classes:
> 
> ```
> QCalendar
> QConcatenateTablesProxyModel
> QSharedMemory
> QTransposeProxyModel
> ```
> 
> I am not sure if this is a bug or not. There is a report that [the shared memory support is disabled when cross-compiling with MinGW](https://bugreports.qt.io/browse/QTBUG-24109), which I am not sure is relevant here.
> 
> A workaround would be to get rid of the references to those classes in PyQt5.
> 
> Another would be to prepare a dummy library which PyQt5 is linked against (or is `LD_PRELOAD`ed) that defines those missing symbols.

error in my termux as follows. 
```
QtCore: 
    QSystemSemaphore, 
    QSharedMemory
QtQuick:
    QSGGeometry
```
take a look at `/data/data/com.termux/files/usr/include/QtCore/qtcore-config.h` and see what features are disabled

**Reply by Mame29** [reply]
how to disable feature QSharedMemory in PyQt5 ?

**Reply by xtkoba** [reply]
Supposedly

```bash
sed -i 's/{.*QSharedMemory, -/{0, 0, -/' sip/QtCore/qobject.sip
sed -i '/qsharedmemory/d' sip/QtCore/QtCoremod.sip
```

Also `sip/QtCore/qsharedmemory.sip` may or may not need modification.

**Reply by Mame29** [reply]
> Supposedly
> 
> ```shell
> sed -i 's/{.*QSharedMemory, -/{0, 0, -/' sip/QtCore/qobject.sip
> sed -i '/qsharedmemory/d' sip/QtCore/QtCoremod.sip
> ```
> 
> Also `sip/QtCore/qsharedmemory.sip` may or may not need modification.

it works fine, but there is an error, should the System Semaphore feature be disabled?, and how to disable it?


>>> import QtCore
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: dlopen failed: cannot locate symbol "sipTypeDef_QtCore_QSystemSemaphore" referenced by "/data/data/com.termux/files/home/PyQt5-5.15.6/build/QtCore/QtCore.abi3.so"...

**Reply by Mame29** [reply]
that's a screenshot of vncserver on termux using qterminal

![anu_build](https://user-images.githubusercontent.com/72637907/152094464-4135c2c5-60e1-4718-b713-77f9f5100f42.png)

**Reply by Mame29** [reply]
thank you for helping

**Comment by Mame29** [comment]
ok we managed to build PyQt5 by disabling :
> QtCore:
>> QSharedMemory
>> QSystemSemaphore

> QtQuick
>> all

![Screenshot_20220202-220339](https://user-images.githubusercontent.com/72637907/152179849-d790421b-ba1d-4dc3-8520-9f31fc4fd4f4.png)


![Screenshot_20220202-214520](https://user-images.githubusercontent.com/72637907/152179108-c62eafd2-3871-44e6-ba91-493ed14c79d9.png)

**Comment by sudomain** [comment]
Can someone provide build instructions? I gather that I need an ARM64 device (done) and qmake from `pkg in qt5-qtbase`, but `pip install PyQt5` failed with many errors- Termux ran out of space in its session buffer lol. I can share it as a log if needed.

Are these the directions that I'm supposed to follow?: https://www.riverbankcomputing.com/static/Docs/PyQt5/installation.html#building-and-installing-from-source

**Reply by Mame29** [reply]
disable QtQuick

**Reply by sudomain** [reply]
A bit out of my depth here, sorry. I found `QT_FEATURE_sharedmemory` and `QT_FEATURE_systemsemaphore` and set them to `0` in `/data/data/com.termux/files/usr/include/QtCore/qtcore-config.h` 
1. is this even the correct file? 
2. Do I additionally change `QT_NO_SHAREDMEMORY` or `QT_NO_SYSTEMSEMAPHORE` in the same file?

I was hoping to find an equivalent file `qtquick-config.h` for QtQuick which makes me think I'm not looking at the correct location at all lol. How do I disable QtQuick?

**Reply by Mame29** [reply]
Try 
```sh
sed -i 's/{.*QSharedMemory, -/{0, 0, -/' sip/QtCore/qobject.sip
sed -i '/qsharedmemory/d' sip/QtCore/QtCoremod.sip
```
For disable QtQuick use `--disable-feature QtQuck`

**Comment by codemona** [comment]
Is building PyQt6 similar process?

**Reply by Mame29** [reply]
no, because the Qt6 library is not yet in the termux repository
